composer/loggers/mosaicml_logger.py

```diff
@@ -0,0 +1,164 @@
+
+
+def _send_data(job_id: str, data: JSON):
+    try:
+        response = requests.post(_MOSAICML_LOGGER_URL,
+                                 headers={"X-MosaicML-API-key": os.environ.get(_MOSAICML_API_KEY_ENV, "")},
+                                 json={
+                                     "experimentID": job_id,
+                                     "data": data,
+                                 },
+                                 timeout=5)
+        if response.status_code >= 400:
+            # Terminate the process in the case of a non-transient error
+            log.error("Posting data to MosaicML backend failed with response code "
+                      f"{response.status_code} and message {response.text}.")
+            sys.exit(1)
+    except requests.exceptions.Timeout as e:
+        # Warn the user in the case of a timeout but don't retry for now
+        log.warning(f"MosaicML logger timed out with error {e}. Logs were not sent to the backend.")
+
+
+class MosaicMLLoggerBackend(RankZeroLoggerBackend):
+    """Log to the MosaicML backend.
+
+    Args:
+        job_id (str, optional): The id of the job to write logs for.
+        creds_file (str, optional): A file containing the MosaicML api_key. If not provided
+            will default to the environment variable MOSAIC_API_KEY.
+        flush_every_n_batches (int): Flush the log data buffer every n batches. (default: ``100``)
+        max_logs_in_buffer (int): The maximum number of log entries allowed in the buffer before a
+            forced flush. (default: ``1000``)
+    """
+
+    def __init__(self,
+                 job_id: Optional[str] = None,
+                 creds_file: Optional[str] = None,
+                 flush_every_n_batches: int = 100,
+                 max_logs_in_buffer: int = 1000) -> None:
+
+        super().__init__()
+        self.skip_logging = False
+        self.job_id = job_id if job_id is not None else os.environ.get(_JOB_ID_ENV, None)
+        if self.job_id is None:
+            self.skip_logging = True
+            warnings.warn("No job_id provided to MosaicLoggerBackend. MosaicML logger will be a no-op.")
+
+        if creds_file:
+            with open(creds_file, 'r') as f:
+                os.environ[_MOSAICML_API_KEY_ENV] = str(f.read())
+
+        if os.environ.get(_MOSAICML_API_KEY_ENV, None) is None:
+            self.skip_logging = True
+            warnings.warn(
+                f"No api_key set for environment variable {_MOSAICML_API_KEY_ENV}. MosaicML logger will be a no-op.")
+
+        self.buffered_data = []
+        self.flush_every_n_batches = flush_every_n_batches
+        self.max_logs_in_buffer = max_logs_in_buffer
+
+        self.queue = Queue()
+        self.thread = Thread(target=self._listen_to_queue)
+
+    def _log_metric(self, epoch: int, step: int, log_level: LogLevel, data: TLogData):
+        del log_level  # unused
+
+        if self.skip_logging:
+            return
+
+        log_data = {
```

Comment:
An example call would be (with curl):

```
$ curl -X POST -H "Content-Type:application/json" -H "X-MosaicML-API-key:A106A6B8" https://api.mosaicml.com/v0/log/metric -d '{"experimentID": "79A8D3D2", "runID": "DE23902E", "data": {"acc/val": 0.5, "wall_clock_time": 123.0}}'
```
(Modulo, that the data field can be an array)

So we need to provide both experimentID and runID, and get the epoch and step in the data object.

This leads me to think we should team up and try this out real quick, and ensure that it works.